### ISSUE TITLE
DNS issues kill SvelteKit SSR process with fetch()

### ISSUE THREAD

**Issue by miohtama** [sveltejs/kit#4775]
### Describe the problem

If a `fetch()` DNS resolution fails the SvelteKit server-side rendering process of node-adapter crashes and dies.

It is unusual for a failed `fetch()` to kill a process, and highly undesirable, so I would like to highlight this unexpected behaviour.

Our [node-adapter](https://github.com/tradingstrategy-ai/frontend/) encountered the following error and crashes the whole SSR process when doing SSR:

```
FetchError: request to https://tradingstrategy.ai/api/candles?pair_id=1656735&time_bucket=4h failed, reason: getaddrinfo EAI_AGAIN tradingstrategy.ai
    at ClientRequest.<anonymous> (file:///root/frontend/build/shims-eb608bdd.js:6246:11)
    at ClientRequest.emit (node:events:526:28)
    at TLSSocket.socketErrorListener (node:_http_client:442:9)
    at TLSSocket.emit (node:events:526:28)
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  type: 'system',
  errno: 'EAI_AGAIN',
  code: 'EAI_AGAIN',
  erroredSysCall: 'getaddrinfo'
```

[More information about the getaddrinfo crash issue here](https://github.com/tradingstrategy-ai/frontend/issues/82#issuecomment-1112733513).

### Describe the proposed solution

`getaddrinfo` errors should be catchable and handled like any other `fetch()` exceptions.

### Alternatives considered

Make sure your SSR process is always autorestarting kind. This way it can recover from these errors without manual intervention.

### Importance

nice to have

### Additional Information

Node v16.14.2

```
├── @sveltejs/adapter-node@1.0.0-next.73
├── @sveltejs/adapter-static@1.0.0-next.18
├── @sveltejs/kit@1.0.0-next.308
├── @testing-library/svelte@3.0.3
├── eslint-plugin-svelte3@3.2.1
├── gridjs-svelte@2.1.1
├── jest-transform-svelte@2.1.1
├── prettier-plugin-svelte@2.4.0
├── svelte-check@2.2.6
├── svelte-highlight@5.1.2
├── svelte-intersection-observer@0.8.0
├── svelte-inview@3.0.0
├── svelte-jester@1.8.2
├── svelte-preprocess@4.9.4
├── svelte-spinner@2.0.2
├── svelte-tradingview-widget@2.3.0
├── svelte-web3@2.2.1
├── svelte@3.47.0
├── sveltestrap@4.2.1
```

**Comment by antony** [comment]
Do you have an example of the try catch block you have written which is not catching this FetchError?

The goal of SvelteKit's fetch is to be as close as possible to `fetch`, so adding special behaviour to it for catching errors would be out of the scope of SvelteKit.

**Comment by miohtama** [comment]
> Do you have an example of the try catch block you have written which is not catching this FetchError?

I do not have a repeatable error, but here is the offending bit:

```javascript
<!-- This was run on SSR when a route is rendered -->
<script lang="ts">
        function reload() { 
         // ...
          const encoded = new URLSearchParams(params);
          const candlesApiUrl = `${backendUrl}/candles?${encoded}`;
          const liquidityApiUrl = `${backendUrl}/xyliquidity?${encoded}`;
          const [candleResp, liquidityResp] = await Promise.all([fetch(candlesApiUrl), fetch(liquidityApiUrl)]);
          if(!candleResp.ok) {
              console.error(candleResp);
              return;
          }
          if(!liquidityResp.ok) {
              console.error(liquidityResp);
              return;
          }
         // ...
     }
</script>
```

[The full source code is here](https://github.com/tradingstrategy-ai/frontend/blob/bbb453d6a6e737458a710f1f6b8f5190e2d371e6/src/routes/trading-view/%5Bchain%5D/%5Bexchange%5D/%5Bpair%5D/index.svelte) (too complex to easily digest).

We tried to wrap it in try { } catch but it did not seem to help. In the end, as a workaround, we skipped loading this data on SSR at the first place.

**Comment by antony** [comment]
I would have really needed to see the code *with* the catch.

However, if you say you wrapped it in a try catch and it didn't catch, then I can only imagine that something in JS (not Svelte) is the cause.

Without a reproduction and a repeatable error though, we won't really be able to help you any further.

If this is important for you, I would invest a little time creating something we can use to see this error for ourselves, and we will endeavour to fix it or upstream it to the correct project.

**Comment by kenkunz** [comment]
Hi @antony – I am the engineer who reported this at [tradingstrategy.ai](https://tradingstrategy.ai). I noticed this was sitting open for more info, so I took a fresh look at it.

I tried creating a minimal repro but I was unable to reproduce the issue. I reverted to an earlier version of the [tradingstrategy-ai / frontend](https://github.com/tradingstrategy-ai/frontend) codebase (where we encountered this)… still could not reproduce it.

Trying to figure out what changed… it's possible I was running a different `node` version when this happened. Tried the most likely version I would have been running… still nothing. At the time, I had been working on a major feature branch with some npm dependency changes – maybe something was jacked with my npm install?

I am flummoxed, but happy report that this no longer appears to be an issue.

@miohtama unless you have anything different to report, I think they can close this issue.

**Comment by sloan-dog** [comment]
@kenkunz I suspect this might be an 'unhandledrejection' termination. I've been seeing this myself, in both node v16.x and 18.x. I tried to resolve with a try catch block, but it doesn't appear to help. I had to add a handler to node's system event 'unhandledrejection' to mitigate node terminating the process. So maybe, there's some side effect or tertiary promise flow that causes the rejected promise not to bubble up properly.
